cpp/src/lib.rs

```diff
@@ -109,3 +109,56 @@ macro_rules! cpp {
         }
     };
 }
+
+#[doc(hidden)]
+pub trait CppTrait {
+    type BaseType;
+    const ARRAY_SIZE: usize;
+    const CPP_TYPE: &'static str;
+}
+
+/// This macro allow to wrap a relocatable C++ struct or class that might
+/// have destructor or copy constructor, and instantiate the Drop and Clone
+/// trait appropriately.
+///
+/// ```ignore
+/// cpp_class!(pub struct MyClass, "MyClass");
+/// impl MyClass {
+///     fn new() -> Self {
+///         unsafe { cpp!([] -> MyClass as "MyClass" { return MyClass(); }) }
+///     }
+///     fn member_function(&self, param : i32) -> i32 {
+///         unsafe { cpp!([self as "const MyClass*", param as "int"] -> i32 as "int" {
+///             return self->member_function(param);
+///         }) }
+///     }
+/// }
+/// ```
+///
+/// This will create a rust struct MyClass, which has the same size and
+/// alignement as the the C++ class "MyClass". It will also implement the Drop trait
+/// calling the destructor, the Clone trait calling the copy constructor, if the
+/// class is copyable (or Copy if it is trivialy copyable), and Default if the class
+/// is default constructible
+///
+/// Warning: This only work if the C++ class can be moved in memory (using
+/// memcpy). This disallow most classes from the standard library.
+///
+#[macro_export]
+macro_rules! cpp_class {
+    (struct $name:ident, $type:expr) => {
```

Comment:
Thats a good idea.

```
cpp! { struct Foo as "Foo" }
```

Note that is was thinking the macro could be also extended to more complex constructs

```
// This will automatically use operator==
cpp! { #[derive(Eq)] struct Foo as "Foo" }

// map the members to the C++ structure and somehow check it is valid
cpp! { struct Point as "Point" {  x : u32,   y: u32  }  }
```

Reply:
Actually, i'm a bit unsure.

having only one macro cpp! could be nice, and would avoid wondering if cpp_class is the right name or if it should be cpp_struct.

However, having a separate macro makes it easier to document or to lookup what it is, than having a single macro that does everything.